File: app/src/xg/encode.ts
// Port of training/encode.py — feature row -> numeric model input, plus the bucket masking.

import type { Bucket, FeatureRow } from "./types";
import {
  ASSIST_TYPES, BODY_PARTS, FEATURE_COLUMNS, FEATURE_GROUPS, PLAY_PATTERNS,
  SHOT_TYPES, TECHNIQUES,
} from "./features";

const CATEGORICALS = ["body_part", "shot_type", "play_pattern", "technique", "assist_type"] as const;

export const NUMERIC_COLUMNS: string[] = FEATURE_COLUMNS.filter(
  (c) => !(CATEGORICALS as readonly string[]).includes(c),
);

const ONEHOT: [string, readonly string[]][] = [
  ["body_part", BODY_PARTS],
  ["shot_type", SHOT_TYPES],
  ["play_pattern", PLAY_PATTERNS],
  ["technique", TECHNIQUES],
  ["assist_type", ASSIST_TYPES],
];

export const ONEHOT_COLUMNS: string[] = ONEHOT.flatMap(([c, vocab]) =>
  vocab.map((v) => `${c}=${v}`),
);

export const MODEL_FEATURES: string[] = [...NUMERIC_COLUMNS, ...ONEHOT_COLUMNS];

/** Which optional groups each completeness bucket keeps (must match train.py BUCKETS). */
export const BUCKET_KEEP: Record<Bucket, Set<string>> = {
  minimal: new Set(),
  partial: new Set(["freeze_frame"]),
  full: new Set(["freeze_frame", "assist", "technique", "pass_origin", "context"]),
};

/** Blank every optional group not in `keep` (categoricals -> null, numerics -> NaN). */
export function maskGroups(row: FeatureRow, keep: Set<string>): FeatureRow {
  const out: FeatureRow = { ...row };
  for (const g of Object.keys(FEATURE_GROUPS)) {
    if (g === "core" || keep.has(g)) continue;
    for (const col of FEATURE_GROUPS[g]) {
      out[col] = (CATEGORICALS as readonly string[]).includes(col) ? null : NaN;
    }
  }
  return out;
}

/** FeatureRow -> Float32Array in MODEL_FEATURES order. Missing -> NaN (numeric) / all-zero (one-hot). */
export function encodeRow(row: FeatureRow): Float32Array {
  const out = new Float32Array(MODEL_FEATURES.length);
  let i = 0;
  for (const c of NUMERIC_COLUMNS) {
    const v = row[c];
    out[i++] = v === null || v === undefined ? NaN : Number(v);
  }
  for (const [col, vocab] of ONEHOT) {
    const cat = row[col];
    for (const v of vocab) out[i++] = cat === v ? 1 : 0;
  }
  return out;
}

File: training/encode.py
"""
encode.py — turn the feature table into a purely numeric model matrix.

We one-hot the five categoricals into fixed columns instead of using LightGBM's native
categorical splits. Reasons:
  * ONNX export of categorical splits is fragile; an all-numeric model converts cleanly,
  * the browser port becomes "set one dummy to 1" with no code tables,
  * a missing category (group dropped) is simply "all dummies 0" — an unambiguous "unknown".

MODEL_FEATURES is the exact input order the ONNX model expects. The TS port must build the
same vector in the same order.
"""

from __future__ import annotations

import numpy as np
import pandas as pd

import features as F

_ONEHOT = [
    ("body_part", F.BODY_PARTS),
    ("shot_type", F.SHOT_TYPES),
    ("play_pattern", F.PLAY_PATTERNS),
    ("technique", F.TECHNIQUES),
    ("assist_type", F.ASSIST_TYPES),
]

ONEHOT_COLUMNS: list[str] = [f"{col}={v}" for col, vocab in _ONEHOT for v in vocab]
MODEL_FEATURES: list[str] = list(F.NUMERIC_COLUMNS) + ONEHOT_COLUMNS

# monotone constraint per MODEL_FEATURES entry: xG falls as it gets harder, rises as easier
_MONO_NUMERIC = {
    "distance": -1, "dist_x": -1, "angle": +1,
    "defenders_in_cone": -1, "nearest_def_dist": +1, "def_within_3": -1, "def_within_5": -1,
    "gk_in_cone": -1, "one_on_one": +1, "open_goal": +1,
}
MONOTONE = [_MONO_NUMERIC.get(c, 0) for c in F.NUMERIC_COLUMNS] + [0] * len(ONEHOT_COLUMNS)


def encode_matrix(df: pd.DataFrame) -> np.ndarray:
    """DataFrame -> float32 array of shape (n, len(MODEL_FEATURES)). NaN kept for missing."""
    parts = [df[F.NUMERIC_COLUMNS].to_numpy(dtype="float32")]
    for col, vocab in _ONEHOT:
        codes = df[col].astype("object").to_numpy()
        oh = np.zeros((len(df), len(vocab)), dtype="float32")
        for j, v in enumerate(vocab):
            oh[:, j] = (codes == v)
        parts.append(oh)
    return np.hstack(parts)
